### PR TITLE
Update cloneWorld().

### PR DESCRIPTION
**Pull request by JBYoshi**
- Make it actually work.
- Support for unloaded worlds - saves memory!
- No longer needs to unload the source world.

Fixes Multiverse/Multiverse-Core#1436
Fixes Multiverse/Multiverse-Core#1491

### CODE REVIEW

File: src/main/java/com/onarandombox/MultiverseCore/utils/WorldManager.java [hunk 1]
```diff
@@ -113,11 +109,31 @@ public boolean cloneWorld(String oldName, String newName, String generator) {
             return false;
         }
 
-        unloadWorld(oldName);
+        boolean needsLoading = this.plugin.getServer().getWorld(oldName) == null;
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseCore/utils/WorldManager.java]: Why is the world being loaded?

**Comment by Superkaj** [comment]
I have no idea on how to import this to eclipse, is it possible for you to create the plugin for itself for us noobs to get a quick fix? :)

**Comment by JBYoshi** [comment]
@dumptruckman Status?

**Comment by dumptruckman** [comment]
I need to fix the build issues caused by dependencies. I've been extremely busy lately. I might be able to check it out next week.

**Comment by JBYoshi** [comment]
@Superkaj I published a download at https://github.com/JBYoshi/Multiverse-Core/releases/tag/JBYoshi-pr1615.

**Comment by Superkaj** [comment]
@JBYoshi Awesome, thanks! :D

**Comment by JBYoshi** [comment]
@dumptruckman I added another API method which automatically computes the generator plugin to use. I also improved error logging.

@Superkaj I'll update the downloadable plugin version later today.

**Comment by JBYoshi** [comment]
@Superkaj Updated: https://github.com/JBYoshi/Multiverse-Core/releases/tag/JBYoshi-pr1615-b2

**Comment by Rigby90** [comment]
Not that it matters too much but are you able to rebase this to include the travis-ci change?

**Comment by dumptruckman** [comment]
I think I fixed the issue with Travis CI. Please rebase and we will review this PR for merge approval.

**Comment by JBYoshi** [comment]
@dumptruckman Looks like it works now. Ready for merge.

File: src/main/java/com/onarandombox/MultiverseCore/utils/WorldManager.java [hunk 2]
```diff
@@ -110,52 +121,47 @@ public boolean cloneWorld(String oldName, String newName, String generator) {
...
+            Logging.warning("File for new world '%s' already exists", newName);
             return false;
         }
 
-        unloadWorld(oldName);
+        MVWorld oldWorld = (MVWorld) this.getMVWorld(oldName);
+        Environment environment = oldWorld.getEnvironment();
+        String seedString = oldWorld.getSeed() + "";
+        WorldType worldType = oldWorld.getWorldType();
+        Boolean generateStructures = null; // TODO actually get a value for this?
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseCore/utils/WorldManager.java]: Does this work as null?

Comment (JBYoshi) [inline src/main/java/com/onarandombox/MultiverseCore/utils/WorldManager.java]: Yes. However, I'll probably change this to grab a value from the world if loaded. Would it be possible to cache it to add support for unloaded worlds as well?

**Comment by JBYoshi** [comment]
Updated download - https://github.com/JBYoshi/Multiverse-Core/releases/tag/JBYoshi-pr1615-b3

@dumptruckman Would it be possible to cache the value of `generateStructures` so `cloneWorld()` can provide the real value for it, even if the world is unloaded? If so, how?

**Comment by JBYoshi** [comment]
@dumptruckman I've finished off everything on this PR. I've tested it on my private server, and this works properly.

Due to some limitations, I have to temporarily load the old world in order to properly clone. However, **no chunks are loaded during this process.** Just the world metadata (such as `level.dat`). The data is also unloaded immediately after `cloneWorld` grabs a small amount of data from it.

@Superkaj Download is at https://github.com/JBYoshi/Multiverse-Core/releases/tag/JBYoshi-pr1615-b4.

**Comment by dumptruckman** [comment]
Thank you JBYoshi. I squashed all these commits and pushed it to master.

**Comment by JBYoshi** [comment]
Thanks!

**Pull request closed without merging**